compiler/qsc_qasm3/README.md

```diff
@@ -0,0 +1,41 @@
+# Q# QASM3 Compiler
+
+This crate implements a semantic transformation from OpenQASM 3 to Q#. At a high level it parses the OpenQASM program (and all includes) into an AST. Once this AST is parsed, it is compiled into Q#'s AST.
+
+Once the compiler gets to the AST phase, it no longer cares that it is processing Q#. At this point all input is indistinguishable from having been given Q# as input. This allows us to leverage capability analysis, runtime targeting, residual computation, partial evaluation, and code generation to the input.
+
+## Process Overview
+
+The OpenQASM code is parsed with their own lexer/parser and any errors are returned immediately before we get into further compilation. The OpenQASM parsing library hard-codes the file system into the types and parsing. Additionally, the library panics instead of surfacing `Result`. Because of this, there is a custom layer built to do the parsing so that the file system is abstracted with the `SourceResolver` trait and results with detailed errors are propagated instead of crashing.
+
+While it would be nice to use their semantic library to analyze and type check the code, they are missing many language features that are only available in the AST and the semantic library panics instead of pushing errors.
+
+With the source lexed and parsed, we can begin compilation. The program is compiled to the Q# AST. The OpenQASM ASG would be great to use for program as a validation pass once it is more developed, but for now we try to surface OpenQASM semantic/type errors as we encounter them. It is difficult to determine if a type error is from the AST compilation or from the source program as the implicit/explicit casting and type promotion rules are very complicated.
+
+As we process the AST we map the spans of the OpenQASM statements/expressions/literals to their corresponding Q# AST nodes. Additionally, we map the input source code (and it’s loaded externals) into a SourceMap with entry expressions. At this point we have enough information to hand off the rest of compilation to the Q# compiler.
+
+## Semantics
+
+The two languages have many differences, and insofar as possible, the compilation preserves the source language semantics.
+
+- OpenQASM and Q# have different qubit and initialization semantics.
+  - Q# assumes that qubits are in the |0⟩ state when they are allocated.
+  - OpenQASM does not make this assumption and qubits are in an undefined state.
+- Q# requires that qubits are reset to the |0⟩ at the end of excution.
+- OpenQASM does not require this.
```

Comment:
nitpick: these feel related in the same way the two above are, so it might make sense to put them together. Also, the restriction on qubit reset is at release, not necessarily at end of execution, but in this context where OpenQASM always allocates qubits at the global scope the two happen to be the same, so not sure if that is a distinction worth highlighting here. As a possible suggestion:

```suggestion
- OpenQASM and Q# have different qubit management semantics.
  - Q# assumes that qubits are in the |0⟩ state when they are allocated.
  - OpenQASM does not make this assumption and qubits start in an undefined state.
  - Q# requires that qubits are reset to the |0⟩ state when released.
  - OpenQASM does not require qubits to be in a specific state at the end of execution.
```